pallets/automation-time/src/migrations.rs

```diff
@@ -1,72 +1,212 @@
 	use super::*;
 
 	pub fn migrate<T: Config>() -> Weight {
-		info!(target: "automation-time", "Migrating automation-time v2");
+		log::info!(target: "automation-time", "Migrating automation-time v3");
 		let pallet_prefix: &[u8] = b"AutomationTime";
 
-		let scheduled_tasks_prefix: &[u8] = b"ScheduledTasks";
-		let _scheduled_tasks: Vec<_> = storage_key_iter::<
-			u64,
-			BoundedVec<T::Hash, T::MaxTasksPerSlot>,
-			Twox64Concat,
-		>(pallet_prefix, scheduled_tasks_prefix)
-		.drain()
-		.collect();
-		let empty_task_queue: Vec<T::Hash> = Vec::new();
-		TaskQueue::<T>::put(empty_task_queue);
-		let empty_missed_queue: Vec<MissedTask<T>> = Vec::new();
-		MissedQueue::<T>::put(empty_missed_queue);
-		let tasks_prefix: &[u8] = b"Tasks";
-		let _tasks: Vec<_> =
-			storage_key_iter::<T::Hash, Task<T>, Twox64Concat>(pallet_prefix, tasks_prefix)
+		// Move all tasks from Tasks to AccountTasks
+		let old_tasks_prefix: &[u8] = b"Tasks";
+		let old_tasks =
+			storage_key_iter::<T::Hash, Task<T>, Twox64Concat>(pallet_prefix, old_tasks_prefix)
 				.drain()
-				.collect();
+				.collect::<btree_map::BTreeMap<T::Hash, Task<T>>>();
+		old_tasks.iter().for_each(|(task_id, task)| {
+			AccountTasks::<T>::insert(task.owner_id.clone(), task_id, task.clone());
+		});
+		let task_migrated = old_tasks.len() as u64;
 
-		let pallet_prefix: &[u8] = b"AutomationTime";
-		let storage_item_prefix: &[u8] = b"LastTimeSlot";
-		let stored_min_slot = get_storage_value::<u64>(pallet_prefix, storage_item_prefix, &[])
-			.expect("Must have last slot value");
-		let diff_to_hour = stored_min_slot % 3600;
-		let current_hour_slot = stored_min_slot.saturating_sub(diff_to_hour);
-		LastTimeSlot::<T>::put((current_hour_slot, current_hour_slot));
-
-		info!(target: "automation-time", "Completed automation-time migration to v2");
-		StorageVersion::new(2).put::<Pallet<T>>();
-		T::DbWeight::get().reads_writes(4, 4)
+		// Move all tasks from ScheduledTasks to ScheduledTasks2
+		let mut times_migrated = 0u64;
+		let old_scheduled_prefix: &[u8] = b"ScheduledTasks";
+		storage_key_iter::<u64, BoundedVec<T::Hash, T::MaxTasksPerSlot>, Twox64Concat>(
+			pallet_prefix,
+			old_scheduled_prefix,
+		)
+		.drain()
+		.for_each(|(time, task_ids)| {
+			let new_task_ids = task_ids
+				.into_iter()
+				.filter_map(|task_id| {
+					if let Some(task) = old_tasks.get(&task_id) {
+						Some((task.owner_id.clone(), task_id))
+					} else {
+						log::debug!(target: "automation-time", "Unable to get task with id {:?}", task_id);
+						None
+					}
+				})
+				.collect::<Vec<_>>();
+
+			let account_task_ids: BoundedVec<AccountTaskId<T>, T::MaxTasksPerSlot> =
+				new_task_ids.try_into().unwrap();
+			ScheduledTasks2::<T>::insert(time, account_task_ids);
+			times_migrated += 1;
+		});
+
+		// Move all tasks from TaskQueue to TaskQueue2
+		let old_task_queue_prefix: &[u8] = b"TaskQueue";
+		let new_task_ids =
+			get_storage_value::<Vec<TaskId<T>>>(pallet_prefix, old_task_queue_prefix, &[])
```

Comment:
Looks like using `take_storage_value` might allow you to avoid calling `kill()` later. (I haven't verified that they're equivalent, but seem to be based on the docs)

The advantage I see here would be that you could delete all of the old storage types from the `lib.rs` file. Deleting the old storage types would remove dead code and prove that you didn't miss any code paths that are accidentally using the old storage still.

The extreme version of this would be to continue using the old (v1) storage prefixes and just modifying them in place, but that might be a bridge too far. It would reduce the size of this change set but it might make things more complex than necessary.

Reply:
ya I had been debating just removing the storage in this PR. For some reason I was hesitant to do it but I am fine making that change. 

I know I can use the same storage and just mutate the contents but I think new storage is safer

Reply:
> I know I can use the same storage and just mutate the contents but I think new storage is safer

Fine by me. I didn't actually know that you could migrate the storage in place. Learned that reading your pr, so thanks!

Reply:
I am going to circle back to this after I test try-runtime. I want to make sure I don't need them for a two part migration